client/src/main/scala/io/delta/sharing/client/DeltaSharingClient.scala

```diff
@@ -719,6 +795,77 @@ class DeltaSharingRestClient(
     )
   }
 
+  private def getQueryInfoTargetUrl(table: Table, queryId: String) = {
+    val shareName = URLEncoder.encode(table.share, "UTF-8")
+    val schemaName = URLEncoder.encode(table.schema, "UTF-8")
+    val tableName = URLEncoder.encode(table.name, "UTF-8")
+    val encodedQueryId = URLEncoder.encode(queryId, "UTF-8")
+    val target = getTargetUrl(
+      s"/shares/$shareName/schemas/$schemaName/tables/$tableName/queries/$encodedQueryId")
+    target
+  }
+
+  private def getQueryInfo(
+              table: Table,
+              queryId: String,
+              maxFiles: Option[Int],
+              pageToken: Option[String]): (Long, String, Seq[String]) = {
+    val target: String = getQueryInfoTargetUrl(table, queryId)
+    val request: GetQueryInfoRequest = GetQueryInfoRequest(
+      queryId = queryId,
+      maxFiles = maxFiles,
+      pageToken = pageToken)
+
+    getNDJson(target, request)
+  }
+
+  private def checkQueryPending(lines: Seq[String]): (Seq[String], Option[String], Boolean) = {
+    val queryStatus = JsonUtils.fromJson[SingleAction](lines(0)).queryStatus
+
+    if(queryStatus == null) {
+        (lines, None, false)
+      } else {
+      (lines.drop(1), Some(queryStatus.queryId), queryStatus.status == QUERY_PENDING_TRUE)
+    }
+  }
+
+  private def getNDJsonWithAsync(
+                table: Table,
+                target: String,
+                request: QueryTableRequest): (Long, String, Seq[String], Option[String]) = {
+    // Initial query to get NDJson data
+    val (initialVersion, initialRespondedFormat, initialLines) = getNDJson(target, request)
+
+    // Check if the query is still pending
+    var (lines, queryIdOpt, queryPending) = checkQueryPending(initialLines)
+
+    var version = initialVersion
+    var respondedFormat = initialRespondedFormat
+
+    val startTime = System.currentTimeMillis()
+    // Loop while the query is pending
+    while (queryPending) {
+      if (System.currentTimeMillis() - startTime > asyncQueryMaxDuration) {
+        throw new IllegalStateException("Query is still pending after " +
+          s"${asyncQueryMaxDuration} ms. Please try again later.")
+      }
+    Thread.sleep(asyncQueryPollIntervalMillis)
+
+      val queryId = queryIdOpt.get
+      val (currentVersion, currentRespondedFormat, currentLines)
+        = getQueryInfo(table, queryId, request.maxFiles, request.pageToken)
+      val (newLines, _, newQueryPending) = checkQueryPending(currentLines)
```

Comment:
Question: will the server send the queryId back?
We should perhaps check that if it does, then the queryId must match.